arch/RISCV/RISCVInstPrinter.c

```diff
@@ -0,0 +1,183 @@
+void RISCV_printInst(MCInst *MI, SStream *O, void *info) 
+{
+  	MCRegisterInfo *MRI = (MCRegisterInfo *) info;
+  	//bool Res = false;
+  	MCInst *NewMI = MI;
+  	// TODO: RISCV compressd instructions.
+  	//MCInst UncompressedMI;
+  	//if (!NoAliases)
+    	//Res = uncompressInst(UncompressedMI, *MI, MRI, STI);
+  	//if (Res)
+    	//NewMI = const_cast<MCInst *>(&UncompressedMI);
+  	if (/*NoAliases ||*/ !printAliasInstr(NewMI, O, info))
+    		printInstruction(NewMI, O, MRI);
+  		//printAnnotation(O, Annot);
+}
+
+static void printRegName(SStream *OS, unsigned RegNo) 
+{
+  	SStream_concat0(OS, getRegisterName(RegNo, RISCV_ABIRegAltName));
+}
+
+/**
+void RISCVInstPrinter::printOperand(const MCInst *MI, unsigned OpNo,
+                                    raw_ostream &O, const char *Modifier) 
+*/
+static void printOperand(MCInst *MI, unsigned OpNo, SStream *O) 
+{
+  	unsigned reg;
+  	int64_t Imm = 0;
+
+  	MCOperand *MO = MCInst_getOperand(MI, OpNo);
+  
+  	//
+  	if (MCOperand_isReg(MO)) {
+    		reg = MCOperand_getReg(MO);
+    		printRegName(O, reg);
+    		if (MI->csh->detail) {
+      			MI->flat_insn->detail->riscv.operands[MI->flat_insn->detail->riscv.op_count].type = RISCV_OP_REG;
+      			MI->flat_insn->detail->riscv.operands[MI->flat_insn->detail->riscv.op_count].reg = reg;
+      			MI->flat_insn->detail->riscv.op_count++;
+    		}
+
+    		return;
+  	}
+
+  	//
+  	if (MCOperand_isImm(MO)) {
+    		Imm = MCOperand_getImm(MO);
+    		if (Imm >= 0) {
+      			if (Imm > HEX_THRESHOLD)
+        			SStream_concat(O, "0x%" PRIx64, Imm);
+      			else
+				SStream_concat(O, "%" PRIu64, Imm);
+    		} else {
+      			if (Imm < -HEX_THRESHOLD)
+				SStream_concat(O, "-0x%" PRIx64, -Imm);
+      			else
+				SStream_concat(O, "-%" PRIu64, -Imm);
+    		}
+
+    		if (MI->csh->detail) {
+      			MI->flat_insn->detail->riscv.operands[MI->flat_insn->detail->riscv.op_count].type = RISCV_OP_IMM;
+      			MI->flat_insn->detail->riscv.operands[MI->flat_insn->detail->riscv.op_count].imm = Imm;
+      			MI->flat_insn->detail->riscv.op_count++;
+    		}
+    		return;
+  	}
+
+  	//assert(MO.isExpr() && "Unknown operand kind in printOperand");
+  	assert(0 && "Unknown operand kind in printOperand");
+  	//MO.getExpr()->print(O, &MAI);
+}
+
+static void printCSRSystemRegister(const MCInst *MI, unsigned OpNo,
+                                   //const MCSubtargetInfo &STI,
+                                   SStream *O) 
+{
+  	assert (0 && "CSR system register hav't support.");
+#if 0
+  unsigned Imm = MI->getOperand(OpNo).getImm();
+  auto SysReg = RISCVSysReg::lookupSysRegByEncoding(Imm);
+  if (SysReg && SysReg->haveRequiredFeatures(STI.getFeatureBits()))
+    O << SysReg->Name;
+  else
+    O << Imm;
+#endif
+}
+
+static void printFenceArg(MCInst *MI, unsigned OpNo, SStream *O) 
+{
+  	unsigned FenceArg = MCOperand_getImm(MCInst_getOperand(MI, OpNo));
+  	assert (((FenceArg >> 4) == 0) && "Invalid immediate in printFenceArg");
```

Comment:
i think we should replace all the assert code. in case we want to handle invalid input, we should return error code, and propagate error upstream.

Reply:
All of the assert() comes from LLVM, I replaced the llvm_unreachable() with assert().
Should I comment it makes a feature work mark, or just delete?

Reply:
you can just comment it out. not the best solution, but i guess we can live with that.